cvise/tests/test_lines.py

```diff
@@ -37,49 +37,238 @@ def advance_until(pass_, state, input_path, predicate):
     write_file(input_path, 'int f() {\nchar x;\n}\nnamespace foo\n{\n}\n')
     p, state = init_pass('0', tmp_path, input_path)
     all_transforms = collect_all_transforms(p, state, input_path)
+
     # removal of the namespace
     assert 'int f() {\nchar x;\n}\n' in all_transforms
     # removal of f()
     assert '\nnamespace foo\n{\n}\n' in all_transforms
+    # check no transform violates curly brace sequences
+    for s in all_transforms:
+        assert is_valid_brace_sequence(s)
 
 
 def test_func_namespace_level1(tmp_path, input_path):
     """Test that arg=1 deletes code inside top-level functions and namespaces."""
     write_file(input_path, 'int f() {\nchar x;\n}\nnamespace foo\n{\nvoid g() {\n}\n}\n')
     p, state = init_pass('1', tmp_path, input_path)
     all_transforms = collect_all_transforms(p, state, input_path)
+
     # removal of code inside f()
     assert 'int f() {\n}\nnamespace foo\n{\nvoid g() {\n}\n}\n' in all_transforms
     # removal of code inside foo
     assert 'int f() {\nchar x;\n}\nnamespace foo\n{\n}\n' in all_transforms
+    # removal of both
+    assert 'int f() {\n}\nnamespace foo\n{\n}\n' in all_transforms
+    # check no transform violates curly brace sequences
+    for s in all_transforms:
+        assert is_valid_brace_sequence(s)
 
 
 def test_multiline_func_signature_level0(tmp_path, input_path):
     """Test that arg=0 deletes a top-level function despite line breaks in the signature."""
     write_file(input_path, 'template <class T>\nSomeVeryLongType\nf()\n{\n}')
     p, state = init_pass('0', tmp_path, input_path)
     all_transforms = collect_all_transforms(p, state, input_path)
+
     assert '' in all_transforms
     # no attempts to partially remove the function
     assert len(all_transforms) == 1
+    # check no transform violates curly brace sequences
+    for s in all_transforms:
+        assert is_valid_brace_sequence(s)
 
 
 def test_multiline_func_signature_level1(tmp_path, input_path):
     """Test that arg=1 deletes a nested function despite line breaks in the signature."""
     write_file(input_path, 'namespace {\ntemplate <class T>\nint\nf()\n{\n}\n}\n')
     p, state = init_pass('1', tmp_path, input_path)
     all_transforms = collect_all_transforms(p, state, input_path)
+
     assert 'namespace {\n}\n' in all_transforms
     # the (multi-line) func must be deleted as a whole, not partially
     # FIXME: Improve the heuristic to not try removing just the opening `namespace {` part,
     # and replace the assertion here with "len(all_transforms) == 1".
     for s in all_transforms:
         assert ('template' in s) == ('f()' in s)
+    # check no transform violates curly brace sequences
+    for s in all_transforms:
+        assert is_valid_brace_sequence(s)
+
+
+def test_class_with_methods_level0(tmp_path, input_path):
+    """Test that arg=0 deletes the whole class definition."""
+    write_file(
+        input_path,
+        """
```

Comment:
How do you find this format of tests? Now that the new hint-based implementation stops juggling whitespaces, we can rewrite all tests in this style.